File: java/p2/src/p2/core/System.java
package p2.core;

import java.net.URL;

import p2.types.basic.TupleSet;
import p2.types.exception.UpdateException;
import p2.types.table.TableName;

public interface System {
	/**
	 * Initialize this instance of the P2 runtime.
	 * 
	 * @param port The network port to use
	 */
	void bootstrap(int port);

	/**
	 * Get the system clock.
	 * @return The current system (logical) clock
	 */
	Clock clock();
	
	/**
	 * Install program file into runtime.
	 * @param owner Program owner.
	 * @param file File containing program text.
	 * @throws UpdateException compilation/installation problems
	 */
	void install(String owner, URL file) throws UpdateException;

	/**
	 * Uninstall program.
	 * @param program Program name
	 * @throws UpdateException Uninstall problems
	 */
	void uninstall(String program) throws UpdateException;

	/**
	 * Schedule a set of tuples to be evaluated by a given program.
	 * @param program Program name
	 * @param name Tuple name
	 * @param insertions Set of tuples to insert and evaluate
	 * @param deletions Set of tuples to delete and evaluate
	 * @throws UpdateException
	 */
	void schedule(String program,
				  TableName name,
				  TupleSet insertions,
				  TupleSet deletions) throws UpdateException;
}

File: lincoln/ruby/lib/core/system.rb
require 'lib/types/table/catalog'
require 'lib/exec/query'
require 'lib/core/schedule'
require 'lib/core/driver'
require 'lib/core/clock'
require 'lib/core/periodic'
require 'lib/core/log'
require 'lib/lang/compiler'
require 'open3'

class System
	#  temporary (pa)
	@@periodic = Periodic.new(@schedule)
	@@programs = Hash.new
	@@RUNTIME = "lib/core/runtime.olg"
		
	def init
		Table.init
		@@query      = QueryTable.new
		@@compile    = CompilerTable.new  ## add back once parser/compiler in place
		@@evaluator  = Driver::Evaluate.new  ## add back once driver in place
		@@schedule   = Schedule.new
		@@clock      = Clock.new("localhost")
		@@periodic   = Periodic.new(@schedule)
		@@log        = Log.new($stderr)
		@@programs   = Hash.new
	end
	
  def System.clock
	  @@clock
  end

  def System.evaluator
	  @@evaluator
  end

	def System.query
	  @@query
  end
  
  def System.periodic
    @@periodic
  end
  
  def System.program(name)
    @@programs.has_key?(name) ? @@programs[name] : nil
  end
	
  def System.install_program(name, program)
    @@programs[name] = program
  end
	
  def install(owner, file)
	##print "\t\tINSTALL #{file}\n"
    compilation = TupleSet.new(@@compile.name)
    compilation << Tuple.new(nil, owner, file, nil)
    schedule(compilation, "runtime")
  end
  
  def schedule(tuples, program)
    @driver.synchronize do
      tmpClass = Class.new(Driver::Task)
      tlam = lambda { return tuples }
      plam = lambda { return program }

      tmpClass.send :define_method, :tuples do
        return tlam.call
      end
      
      tmpClass.send :define_method, :program do
        return plam.call
      end
  
      @driver.task(tmpClass.new)
    end
  end  
  
  def bootstrap
    return if defined? @@initialized
    compiler = Compiler.new("system", @@RUNTIME);
    compiler.program.plan
    @@clock.insert(@@clock.time(0),nil)
			
    @driver = Driver.new(System.program("runtime"), @@schedule, @@periodic, @@clock);
			
    Compiler.files.each {|file| install("system", file)}
    Thread.new { @driver }
    @@initialized = TRUE
	end
	
  def main(args)
    bootstrap
    ##print "BOOTSTRAP done. should have table in scope\n"
    tabn = Table.find_table(TableName.new("runtime","priority"))
    #$catalog.tuples.each do |c|
    #	puts c
    #end
    ##print "FOUND (#{tabn})\n"
    args.each { |a| install("user", a) }
    @driver.run()
  end
end
